servicebindings/resolver.go

```diff
@@ -0,0 +1,241 @@
+package servicebindings
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+)
+
+// Binding represents metadata related to an external service.
+type Binding struct {
+
+	// Name is the name of the binding.
+	Name string
+
+	// Path is the path to the binding directory.
+	Path string
+
+	// Type is the type of the binding.
+	Type string
+
+	// Provider is the provider of the binding.
+	Provider string
+
+	// Entries is the set of entries that make up the binding.
+	Entries map[string]*Entry
+}
+
+// Resolver resolves service bindings according to the kubernetes binding spec:
+// https://github.com/k8s-service-bindings/spec#workload-projection.
+//
+// It also supports backwards compatibility with the legacy service binding spec:
+// https://github.com/buildpacks/spec/blob/main/extensions/bindings.md
+type Resolver struct {
+	bindingRoot string
+	bindings    []Binding
+}
+
+// NewResolver returns a new service binding resolver.
+func NewResolver() *Resolver {
+	return &Resolver{}
+}
+
+// Resolve returns all bindings matching the given type and optional provider (case-insensitive). To match on type only,
+// provider may be an empty string. Returns an error if there are problems loading bindings from the file system.
+//
+// The location of bindings is given by one of the following, in order of precedence:
+//
+//   1. SERVICE_BINDING_ROOT environment variable
+//   2. CNB_BINDINGS environment variable, if above is not set
+//   3. `<platformDir>/bindings`, if both above are not set
+func (r *Resolver) Resolve(typ string, provider string, platformDir string) ([]Binding, error) {
```

Comment:
Did you know you can reduce some of the verbosity in these params by declaring types for sequential params that have that same type?

```suggestion
func (r *Resolver) Resolve(typ, provider, platformDir string) ([]Binding, error) {
```

Reply:
Yeah, I know about that but have never preferred it (as it makes modification cumbersome IMO). Happy to conform with the patterns of this repo though.

Reply:
Done here and elsewhere